### DISCUSSION TITLE
How long takes TornadoVM JIT compiler to transform Java bytecode into OpenCL

### DISCUSSION THREAD

**Discussion by MLobatoR** [beehive-lab/TornadoVM#144] [General]
Hello :)

How long should I wait for TornadoVM JIT compiler to transform Java bytecode into OpenCL? Or if is taking too long should I stop the process? I have no exceptions so far but the process started 5h ago.

**Comment by stratika** [comment]
Hi @MLobatoR, the compilation time depends on the actual architecture you target.

**A)** In case, you compile OpenCL for CPUs or GPUs, this should not take more than few milliseconds.

**B)** In case, you compile for FPGAs, it is common to compile for several hours. To be fair the actual compilation of TornadoVM that translates Java bytecodes to OpenCL will still take a few milliseconds. However, a second compilation takes place from the FPGA HLS which is transparently performed by TornadoVM. We have seen such cases, in which the compilation can take up to 3 hours.

Which one is your case?

**Reply by MLobatoR** [reply]
Hi stratika,

My case is A) and I waited for 6h30min hours before stopping manually. I don't have any errors on my console so I cannot tell if there's a problem or not.

I was expecting that if there was a problem with TornadoVM JIT compiler I would get an error or something. But it just keeps running and I don't know what it's doing.

Thanks in advance for your help.

**Reply by stratika** [reply]
This seems very unfortunate. Usually, if there is a problem in the TornadoVM runtime or the JIT compiler, there will be an exception with a description, or in the worst case it could be a NULL pointer exception.

But having the process hanging seems odd. Have you tried running the unit-tests, just to be sure that TornadoVM has been installed properly and it can run on your machine. Can you provide a description of your machine?

To run unit-tests, you can type:
```
make tests
```

**Reply by jjfumero** [reply]
Hi @MLobatoR, 
    it is hard to reproduce the error without more details. My guess is that the TornadoVM JIT compiler generates code for GPU and the code is being executed in an infinite loop. Is it possible for you to share a test-case? 

Also, you can check if the JIT compiler finished by using the following flags:

```bash
tornado --threadInfo --printKernel --debug <yourApp> 
```
If you see the OpenCL code and information about the thread deployment on the GPU, that means that the GPU code is actually running. 

Let us know if this helps.

**Reply by MLobatoR** [reply]
Hi,

@stratika I did run the tests and apart some that are "NOT SUPPORTED", don't know why, the rest passed.

@jjfumero I have the code running without TornadoVM, with correct results. So I supposed we can discard any idea of an infinite loop. I'm running tornado on eclipse IDE and it seems it gets lost somewhere. I can see Tornado threads being created :
![image](https://user-images.githubusercontent.com/60794518/130787775-35ab448a-f1e5-46e3-8ac8-0c5d35ad7ce2.png)

Here is a peace of my code, but I don't think it will help much:
`private void runAlgoForMatrix() {
    	int nbPoints = this.nodes.length;
    	
    	NeoNodeAStarSingleThreadData allData[] = new NeoNodeAStarSingleThreadData[nbPoints];
    	for(int i = 0; i < nbPoints; i++) {
    		allData[i] = new NeoNodeAStarSingleThreadData(this, this.nodes, this.bestWeight, this.weighting, this.pointsNotFound);
    		allData[i].init(i);
    	}
    	
    	System.out.println("Start Task Schedule");
    	
    	TaskSchedule ts = new TaskSchedule("s0");
    	
		TaskSchedule t = ts.task("map", NeoAStarAlgorithm::runParallel, allData, nbPoints, pointsNotFound);
		t.streamOut(this.bestWeight);
    	
    	ts.execute();
    	
    	System.out.println();
    }
    
    public static void runParallel(NeoNodeAStarSingleThreadData allData[], int nbPoints, IntArrayList pointsNotFound) {
    	System.out.println("Start Run Paralell");
    	for(@Parallel int i = 0; i < nbPoints; i++) {
    		if(pointsNotFound.contains(i)) {
    			continue;
    		}
    		
    		System.out.println("Run for " + i);

    		NeoNodeAStarSingleThreadData data = allData[i];

    		boolean finishedFwd = false;
    		while(!finishedFwd) {
    			finishedFwd = data.isFinishedFrom();
    			if(!finishedFwd) {
    				finishedFwd = !data.fillEdgesFrom(false);
    			}

    			data.finishedFrom = finishedFwd;
    		}
    	}
    }`

The last print that I have is "Start Task Schedule".

Are you able to help me?

**Reply by stratika** [reply]
Thanks @MLobatoR.

With a first look, there are two parts that would not work, as they are not supported by TornadoVM. Both of them should result in a compilation error instead of running for hours.

- The first part is the limited support of Objects as parameters to the method that will be compiled for hardware acceleration. The valid parameters can be array of Java primitive types, or some object types (e.g. `VectorFloat`, `VectorFloat4`, etc.) that TornadoVM exposes.

- The second part is that the printing inside the compiled method which is also not supported.

We have a list of what is currently not supported by TornadoVM [here](https://github.com/beehive-lab/TornadoVM/blob/master/assembly/src/docs/Unsupported.md).

I hope this will be useful.

**Reply by MLobatoR** [reply]
@stratika thank you for your input. It was very helpful.

Best regards,
ML

**Reply by MLobatoR** [reply]
Hi @stratika,

I'm trying to use tornado with a whole new algorithm and I'm still struggling with the time this is hanging.

This is what I'm running:
```

	protected static void expand(VectorInt Va, VectorInt Ea, VectorInt pathCost, int[] edgesToExpand, Matrix2DInt results, VectorInt cost, int size) {
		for(@Parallel int i = 0; i < size; i++) {
			int index = edgesToExpand[i];
			KernelGPU2.compute(Va.get(index), Va.size() > index + 1 ? Va.get(index + 1) : Va.size(), pathCost.get(index), Ea, results.row(i), cost);
		}
	}

```


`
public class KernelGPU2 {

	public static void compute(int initIndex, int nextIndex, int initialCost, VectorInt Ea, VectorInt updatedCosts, VectorInt costMatrix) {
		for(int i = initIndex; i < nextIndex; i++) {
			int positionToUpdate = Ea.get(i);
			int currentCost = updatedCosts.get(positionToUpdate);
			int updatedCost = initialCost + costMatrix.get(i);
			if(currentCost > updatedCost) {
				updatedCosts.set(positionToUpdate, updatedCost);
			}
		}
	}
}`

Can you help me? This time I'm using only objects that TornadoVM supports. Also I don't know if I'm compiling for GPU. How can I confirm this?

Thank you very much in advance,
ML

**Reply by stratika** [reply]
Hi @MLobatoR,

Can you provide some information regarding the initialization of the parameters of the `expand` method?
Also, it would be helpful to have the code snippet that creates the `TaskSchedule` in order to reproduce your problem.

**Reply by MLobatoR** [reply]
Hi @stratika,

They are initialized as int[] or int[][] and then added to a VectorInt or into a Matrix2DInt.

And this is how I'm creating TaskSchedule
```
			TaskSchedule ts = new TaskSchedule("s0");
			TaskSchedule t = ts.task("map", NeoAlgorithmGPU2::expand, graph.Va, graph.Ea, pathCost, edgesToExpand, results, graph.meters, edgesToExpand.size());
			t.streamOut(results);
			
			t.execute();
```
Do you need more information?

Thank you,
ML

**Reply by jjfumero** [reply]
@MLobatoR , what is the issue you are getting? is it a compilation error? is it possible for you to send a test-case to reproduce the problem? 
Also, could you please attach the driver and backend you are using?

**Reply by MLobatoR** [reply]
Hi @jjfumero,

I manage to understand what was wrong. It seams like I cannot something like this:

`Va.size() > index + 1 ? Va.get(index + 1) : Va.size()`

in methods that will be called by my task sheduler. It makes sense, and now I don't have any exceptions nor the compilation is hanging.

That said, I'm now struggling with the result. If I turn off parallelization the result is correct, but when I turn parallelization on the result comes out completely wrong. 

So I have the parallelization in a while cycle, something like this:

```
		while(!stop(mark)) {
			VectorInt edgesToExpand = getEdgesToExpand(mark);
			Matrix2DInt results = copyArray(pathCostArray, edgesToExpand.size());
			
			TaskSchedule ts = new TaskSchedule("s"+iteration)
					.task("t"+iteration, NeoAlgorithmGPU2::expand, graph.Va, graph.Ea, pathCost, edgesToExpand, results, graph.meters, edgesToExpand.size())
					.streamOut(results);
			
			ts.execute();
			
			update(pathCostArray, mark, edgesToExpand.size(), results);
		}
```
The first iteration presents the correct result, but second iteration it's all wrong. All the values in results come out wrong. If I comment the task schedule and call the method expand directly instead results are correct.

Can you help me? Can I run the code like this or the while cycle cannot be done with a task schedule inside?

Thank you in advance,
ML

**Reply by MLobatoR** [reply]
Hi @jjfumero,

Now I think the problem is related with the topic I opened yesterday -> https://github.com/beehive-lab/TornadoVM/discussions/213. After you fix this I'll try again.

Thank you for your help,
ML

**Reply by jjfumero** [reply]
A note from your code snippet:

```java
while(!stop(mark)) {
	VectorInt edgesToExpand = getEdgesToExpand(mark);
	Matrix2DInt results = copyArray(pathCostArray, edgesToExpand.size());
			
	TaskSchedule ts = new TaskSchedule("s"+iteration)
	  .task("t"+iteration, NeoAlgorithmGPU2::expand, graph.Va, graph.Ea, pathCost, edgesToExpand, results, graph.meters, edgesToExpand.size())
	.streamOut(results);
        ts.execute();
			
	update(pathCostArray, mark, edgesToExpand.size(), results);
}
```

This means that for every iteration, you build a new `TaskSchedule` and run it. Note that TornadoVM compiles Java to {OpenCL | SPIR-V or PTX} the first time the task-schedule is executed. You can take advantage of the code-cache by compiling once, assuming you can have constant size across iterations. 

Is it possible something like this (this is just a sketch) ? 

```java
// Create Task-Schedule
VectorInt edgesToExpandWithBigBuffer = getEdgesToExpand(mark);
Matrix2DInt results = copyArray(pathCostArray, edgesToExpand.size());

TaskSchedule ts = new TaskSchedule("s"+iteration)
         .streamIn(pathCostArray, edgesToExpandWithBigBuffer)      // TornadoVM will transfer data every iteration 
	  .task("t"+iteration, NeoAlgorithmGPU2::expand, graph.Va, graph.Ea, pathCost, edgesToExpand, results, graph.meters, edgesToExpandWithBigBuffer.size())
	.streamOut(results);

while(!stop(mark)) {	
        ts.execute();
	update(pathCostArray, mark, edgesToExpand.size(), results);
}
```

**Reply by MLobatoR** [reply]
Hi @jjfumero and @stratika,

It is working now. Thank you so much for your assistance. 

I have one last question, this one related with the new version.
What should be considered as an object to lock in Memory, to transfer for device and transfer for host? What is the criteria I should follow to know which one is which?

Thank you,
ML

**Reply by jjfumero** [reply]
Thanks for the question. We are preparing documentation to be released for 0.15 that explains all these new functions. So please, do not hesitate to ask. We understand some of them can be confusing. Having said that, we are also open to feedback. 

### Short answer 

So the new function `lockObjectsInMemory` is a way in TornadoVM to lock a device buffer and avoid alloc/dealloc every time a task-graph is executed (`taskGraph.execute()`).


###  Long answer

We have changed the way TornadoVM handles memory. Currently, with the latest APIs, TornadoVM will alloc and dealloc device buffers. If a task-graph is going to be executed just once, then this should be the way to go, and it is the default mode in Tornado right now. 

However, if a task-graph is going to be executed multiple times, we can "optimize" buffer allocation by reusing device buffers as much as possible. That's the purpose of the `lockObjectsInMemory` function. This means that the first time the task-graph is executed, TornadoVM will do a malloc for all device buffers. When the `execute` finishes, TornadoVM **will not** release device memory resources. From the second time the task-graph is executed, TornadoVM will directly reuse all device buffers already created. This has a huge impact on performance. 

The reason we do not have documented this function yet is that we are adding a `TornadoVM-Executor`. What we want to achieve is a separation between the task-graph definition and task-graph execution plan and optimization. We are planning to move this function (`lockObjectsInMemory`) to the executor once we have it.

**Reply by MLobatoR** [reply]
@jjfumero thank for your clarification, I think I understand now.

Hope this is the last question :) but if I want to run the same code in my while cicle and the only thing that changes are the input variables is there a way to not be always generating graph?

Thank you very much for all your help and assistance,
ML

**Reply by jjfumero** [reply]
Yes, there are a few ways:

1) If your input/output arrays do not change in length over time, then you can use the following template:

```java
// pseudocode 
TaskGraph tg = new TaskGraph("sample")
            .lockObjectsInMemory(arr1, arr2, output1, output2, .... )      // This is to optimize data allocation between calls
            .transferToDevice(DataTransferMode.EVERY_EXECUTION, arr1, arr2, ...)  // This is to stream-In data from Host -> Device 
            .task( "tk1", ... )
            .task( "tk2", ... ) 
            .transferToHost(output1, output2, ...) ;      // Stream-Out data Device -> Host 

while (condition) {
    tg.execute();

    checkCondition();
}
```

2) If your Input/Output arrays change length over time, you need to use another method called `replaceParameter` every time there is a new resize. This is because TornadoVM, when it compiles the code from Java to {OpenCL | SPIR-V | PTX }, it also performs partial evaluation, and it propagates loop-bounds and some other primitive as constants. 


```java
// pseudocode 
TaskGraph tg = new TaskGraph("sample")
            .transferToDevice(DataTransferMode.EVERY_EXECUTION, arr1, arr2, ...)
            .task( "tk1", ... )
            .task( "tk2", ... ) 
            .transferToHost(output1, output2, ...) ;

tg.execute();

while (condition) {

    tg.execute();
    tg.replaceParameter(oldArray1, newArray1);
    tg.replaceParameter(oldArray2, newArray2); 
    ... 

    checkCondition();
    checkReferences(oldArrays, newArrays);
}
```

You can see some examples here: https://github.com/beehive-lab/TornadoVM/blob/master/tornado-unittests/src/main/java/uk/ac/manchester/tornado/unittests/dynsize/Resize.java#L67

**Reply by MLobatoR** [reply]
@jjfumero that a great help 👍 

Thank you again for all your help,
ML
